nestkernel/eprop_archiving_node.cpp

```diff
@@ -0,0 +1,253 @@
+  if ( it == update_history_.end() or it->t_ != shift )
+    update_history_.insert( it, HistEntryEpropUpdate( shift, 1 ) );
+  else
+    ++it->access_counter_;
+}
+
+void
+nest::EpropArchivingNode::write_update_to_history( long t_previous_update, long t_current_update )
+{
+  long shift = get_shift();
+
+  std::vector< HistEntryEpropUpdate >::iterator it;
+
+  it = std::lower_bound( update_history_.begin(), update_history_.end(), t_current_update + shift );
+
+  if ( it != update_history_.end() or it->t_ == ( t_current_update + shift ) )
+    ++it->access_counter_;
+  else
+    update_history_.insert( it, HistEntryEpropUpdate( t_current_update + shift, 1 ) );
+
+  it = std::lower_bound( update_history_.begin(), update_history_.end(), t_previous_update + shift );
+
+  if ( it != update_history_.end() or it->t_ == ( t_previous_update + shift ) )
+    --it->access_counter_;
+}
+
+void
+nest::EpropArchivingNode::write_eprop_parameter_to_map( std::string parameter_name, double parameter_value )
+{
+  eprop_parameter_map_[ parameter_name ] = parameter_value;
+}
+
+void
+nest::EpropArchivingNode::write_surrogate_gradient_to_history( long time_step, double surrogate_gradient )
+{
+  eprop_history_.push_back( HistEntryEpropArchive( time_step, surrogate_gradient, 0.0 ) );
+}
+
+void
+nest::EpropArchivingNode::write_error_signal_to_history( long time_step, double error_signal )
+{
+  long shift = delay_out_norm_;
+  eprop_history_.push_back( HistEntryEpropArchive( time_step - shift, 0.0, error_signal ) );
+}
+
+void
+nest::EpropArchivingNode::write_learning_signal_to_history( long& time_step,
+  long& delay_out_rec,
+  double& weight,
+  double& error_signal )
+{
+  long shift = delay_rec_out_ + delay_out_norm_ + delay_out_rec;
+
+  std::deque< HistEntryEpropArchive >::iterator it_hist;
+  std::deque< HistEntryEpropArchive >::iterator it_hist_end;
+
+  get_eprop_history( time_step - shift, &it_hist );
+  get_eprop_history( time_step - shift + delay_out_rec, &it_hist_end );
+
+  if ( it_hist != it_hist_end )
+  {
+    it_hist->learning_signal_ += weight * error_signal;
+    ++it_hist;
+  }
+}
+
+void
+nest::EpropArchivingNode::write_firing_rate_reg_to_history( long t_current_update, double f_target, double c_reg )
+{
+  long const update_interval = kernel().simulation_manager.get_eprop_update_interval().get_steps();
+  long shift = Time::get_resolution().get_steps();
+
+  double f_av = n_spikes_ / static_cast< double >( update_interval );
+  double f_target_ = f_target / 1000.0; // convert to kHz
+  double firing_rate_reg = c_reg * ( f_av - f_target_ ) / static_cast< double >( update_interval );
+
+  firing_rate_reg_history_.push_back( HistEntryEpropFiringRateReg( t_current_update + shift, firing_rate_reg ) );
+}
+
+const long
+nest::EpropArchivingNode::get_shift() const
+{
+  long shift_rec = offset_gen_ + delay_in_rec_;
+  long shift_out = offset_gen_ + delay_in_rec_ + delay_rec_out_;
+
+  return get_name() == "eprop_readout" ? shift_out : shift_rec;
+}
```

Comment:
There is no point in making the return value `const`, I suggest to remove that throughout.

More importantly, since this method implements different behavior for different subclasses of `EpropArchivingNode`, it should be virtual with the general case (`return offset_gen_ + delay_in_rec_;`) implemented here. `eprop_readout` can then override the method with its implementation.

Reply:
First part fixed in https://github.com/nest/nest-simulator/pull/2867/commits/87a8062d34b92f78bea73d9fbcba400d3436e7da.

Reply:
Second part fixed in https://github.com/nest/nest-simulator/pull/2867/commits/55913ae0ec3ee9fc2c1292690cd28e016124d156.